compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -611,9 +618,21 @@ impl<'tcx> OrphanChecker<'tcx> {
         ControlFlow::CONTINUE
     }
 
-    fn found_param_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<OrphanCheckEarlyExit<'tcx>> {
+    #[instrument(skip(self), level = "debug")]
+    fn found_uncovered_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<OrphanCheckEarlyExit<'tcx>> {
         if self.search_first_local_ty {
-            ControlFlow::CONTINUE
+            return ControlFlow::CONTINUE;
+        };
+
+        // Fully concrete projections are OK, so check
+        // and only exit if there's generics
+        if let ty::Projection(proj) = *t.kind() {
+            if proj.has_type_flags(TypeFlags::HAS_TY_PARAM) {
+                debug!("found_uncovered_ty: type parameter found in projection, exit orphan check");
+                ControlFlow::Break(OrphanCheckEarlyExit::ParamTy(t))
+            } else {
+                ControlFlow::CONTINUE
+            }
         } else {
```

Comment:
can you instead move that check to the `Projection` branch of the `visit_ty` method?